### PR TITLE
Fix LuceneOnFaiss to have the same cosine score values with Faiss

### PR DESCRIPTION
**Pull request by 0ctopus13prime**
### Description
This PR fixes score translation bug in LuceneOnFaiss when cosine space type is being used.
For more details, please refer to this [issue](https://github.com/opensearch-project/k-NN/issues/2887).

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]
https://github.com/opensearch-project/k-NN/issues/2887

### Check List
- [X] New functionality includes testing.
- [X] New functionality has been documented.
- [X] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [X] Commits are signed per the DCO using `--signoff`.
- [X] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+ * the Faiss score into the maximum inner product range so Lucene can
+ * interpret it correctly during search.</p>
+ *
+ * <p>Conversely, it also converts Lucene scores back into Faiss scores so that
+ * the same query produces consistent results across both implementations.
+ *
+ * <p>Note that this should be used only when memory_optimized_search is enabled.
+ *
+ */
+public final class MemoryOptimizedSearchScoreConverter {
```
Comment (Vikasht34) [inline src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java:37]: This is good approach to fill the gap, So as per current logic , No Cosine Spaces falls back to KNNEngine.LUCENE methods while cosine gets special handling, but when  we have another space type like say Jaccard and if that needs special handling, we would be end up creating similar. Can we have Scorer translator inteface with per space inplementation , we can start with Cosine but it will to extend.

Right now there is already mapping logic leaves in SpaceType and KnnEngine Adapters. Can we not move this special case functions into spaceType having both MOS, NON MOS code at same path?

File: src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        return SpaceType.INNER_PRODUCT.scoreTranslation(-innerProductValue);
+    }
+
+    /**
+     * This method converts Lucene's max inner product score to Faiss cosine score to ensure user
+     * to get the same results with the same query.
+     *
+     * @param scoreDocs Results from internal search before returning.
+     */
+    public static void convertToCosineScore(final ScoreDoc[] scoreDocs) {
```
Comment (Vikasht34) [inline src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java:84]: As part of this , Can we add bound check (eg. [0,1]) like for cosine to catch errors in scores? Another point Since FAISS allows zero vectors [IP=0] but norm =0 causes division by zero in cosine, Can to safe guard , if IP extraction yeilds NAN , can we default to zero. Can we add test cases with Zero Vector.

if(Float.IsNan(scoreDoc.score) || scoreDoc.score <=0 then scoreDoc.score = 0.0f;

File: src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        for (final ScoreDoc scoreDoc : scoreDocs) {
+            // For cosine similarity, MAXIMUM_INNER_PRODUCT being used internally.
+            // Which maps negative values (which is plain inner product result value) to (0, 1], and maps positive values to (1, +inf).
+            // Below logic is to reverse back and extract the result of plain inner product.
+            final float innerProductValue;
+            if (scoreDoc.score > 1) {
```
Comment (Vikasht34) [inline src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java]: I think assumption here is Lucene MAX_IP emitted values strictly > for positive IP and (0,1] for negative IP, But at FP16/Quantized ADC boundry drift may happen, like a vlaue extermly close to 1 , if our boundry condition is strict > 1.0f and <= 1.0f with no tolerance then near boundry samples flip sides. 

FP Math 

fp32 ulp at 1.0 ~~1.1920929e-7 wheere as FP 16 ulp = 0.0009765625, 

if underliying engine accumilates in FP16 then effective uncernatiity around 1.0 is less. 

Can we add some clamping like if (s-1F) <1e-6f consider this as 1F ?

File: src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                // Inner product value is positive.
+                innerProductValue = scoreDoc.score - 1;
+            } else {
+                // Inner product value is negative.
+                innerProductValue = 1 - 1 / scoreDoc.score;
```
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java:95]: if scoreDoc.score = 0 then this code will lead divide by zero exception. can we have a check and a UT for this case.

File: src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java (hunk 1 above)
Comment (0ctopus13prime) [reply src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java:37]: This issue is a bit tricky and comes from a subtle difference in how Lucene and Faiss handle inner product scoring.

Lucene uses MaximumInnerProduct (MIP), which maps negative values to the range (0, 1) and non-negative values (including 0) to [1, +∞). Faiss uses a plain InnerProduct without that mapping.

Since Faiss only supports L2 and InnerProduct, we need special handling in LuceneOnFaiss to reconcile this difference. For the other metrics, we’ve already aligned the behavior, so scores from LuceneOnFaiss can be used interchangeably. **_Cosine is the only exception._**

Also, the SpaceType and KnnEngine adapters aren’t useful here, because they translate engine distances into engine-specific scores. What we need instead is to translate Lucene’s score into Faiss’s score, and only for the cosine case.

That’s why I don’t think this logic belongs in a general interface — it’s a LuceneOnFaiss-specific special case.

File: src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java (hunk 2 above)
Comment (0ctopus13prime) [reply src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java:84]: Hmm… thanks for the suggestion, but I think the score values are guaranteed to be non-NaN.
Also, the translation doesn’t involve division by zero: zero and all positive values are mapped to (value + 1), while only negative numbers go into the denominator.

File: src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java (hunk 3 above)
Comment (0ctopus13prime) [reply src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java]: Small correction is, negative IP will be mapped to (0, 1). If a value that is extremely close to 1, since the value is positive, it will be mapped to 1.99999 something similar. 
Also both Faiss and LuceneOnFaiss first converts FP16 to FP32 then do the distance calculation, I'm skeptical that drift may happen since all score calculation will be conducted in a FP32 variable.

File: src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java (hunk 1 above)
Comment (Vikasht34) [reply src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java:37]: Agree that It's LuceneOnFaiss use case specific , but the root issue is Lucene MIP vs Faiss plain IP and this pattern could re-occure. Lucene's VectorSimiliarityFunction enum already handles similar abstraction and per metrics translations, my worry is this approach would not scale much in future , if we have another score mismatching with lucene and faiss and when we support the Lucene on faiss , we end up creating similar. My Suggestion is see how we can generalize this to have less friction is future and make it state so that in future we don't end up creating this fragmentation in code.

File: src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java (hunk 2 above)
Comment (Vikasht34) [reply src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java:84]: This what I simply tried 

```
import numpy as np

def mip_encode(dot):
    if dot >= 0:
        return 1 + dot
    else:
        return 1.0 / (1 - dot)

def mip_decode(mip):  # Like PR's convertToCosineScore extraction
    if mip > 1:
        return mip - 1  # Positive ip
    else:
        return 1 - 1 / mip  # Negative ip (risky division)

# Case 1: Tiny negative dot (precision loss)
dot = np.float32(-1e-8)  # Tiny dissimilar
mip = mip_encode(dot)
print(f"Dot={dot}, MIP={mip}")  # MIP=1.0 (rounded, lost negative)
ip_decoded = mip_decode(mip)
print(f"Decoded IP={ip_decoded}, Cosine={(1 + ip_decoded)/2}")  # IP=0, Cosine=0.5 (neutral, but should be ~0.499999995)

# Case 2: Overflow to -inf (unnormalized huge negative)
dot = np.float32(-1e38)  # Beyond float32 max (~-3.4e38), becomes -inf
mip = mip_encode(dot)  # 1 / inf = 0.0
print(f"Dot={dot}, MIP={mip}")
ip_decoded = mip_decode(mip)  # 1 - 1/0.0 = -inf
print(f"Decoded IP={ip_decoded}, Cosine={(1 + ip_decoded)/2}")  # Cosine=-inf (bad)

# Case 3: NaN dot (invalid vector)
dot = np.float32(np.nan)
mip = mip_encode(dot)
print(f"Dot={dot}, MIP={mip}")
ip_decoded = mip_decode(mip)
print(f"Decoded IP={ip_decoded}, Cosine={(1 + ip_decoded)/2}")  # All NaN


Output 

1. Case 1: Dot=-1e-08, MIP=1.0
Decoded IP=0.0, Cosine=0.5  (Lost the tiny dissimilarity; should be <0.5)
2. Case 2: Dot=-inf, MIP=0.0
Decoded IP=-inf, Cosine=-inf  (Inf propagation from 1/mip)
3. Case 3: Dot=nan, MIP=nan
Decoded IP=nan, Cosine=nan  (Full propagation)
```

For huge negative dots  float32 can overflow dot to -inf, making (1 - dot)=inf, mip=1/inf=0.0. Decoding then hits 1 / mip = inf, yielding ip=-inf and cosine=NaN/inf

My sugestion is to keep the check.

File: src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java (hunk 1 above)
Comment (0ctopus13prime) [reply src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java:37]: I think it should be fine.
Regarding 

>  if we have another score mismatching with lucene and faiss and when we support the Lucene on faiss , we end up creating similar.

Except for Cosine, all metrics are identical between two, only Cosine is the issue.

File: src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java (hunk 2 above)
Comment (0ctopus13prime) [reply src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java:84]: Please note that Case 2 and Case 3 will not occur. Since we ensure all valid float numbers to be indexed.
And for the Case-1, treating -1e-08 as 0, I think this is reasonable, -1e-08 is almost 0.

Comment (0ctopus13prime) [reply src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java:84]: Also please note that, this PR is to ensure the same score values between Faiss vs LuceneOnFaiss!
When memory optimized search is disabled, it will encode the plain inner product values to max-inner-product and return. Therefore, we need to ensure the same value to be returned. And for the all cases, I got the same results!

```
public class DistanceTest {
    public static void main(String ... args) {
        printResults(-1e-08f);
        printResults(Float.NEGATIVE_INFINITY);
        printResults(Float.NaN);
    }

    public static float scaleMaxInnerProductScore(float vectorDotProductSimilarity) {
        return vectorDotProductSimilarity < 0.0F ? 1.0F / (1.0F + -1.0F * vectorDotProductSimilarity) : vectorDotProductSimilarity + 1.0F;
    }

    public static float cosine(float ip) {
        ip = 1 - ip;
        return Math.max((2.0F - ip) / 2.0F, 0.0F);
    }

    public static void printResults(final float ip) {
        // No MOS
        System.out.println("No MOS (e.g. Faiss C++): final score=" + cosine(ip));

        final float score = scaleMaxInnerProductScore(ip);
        final float innerProductValue;
        if (ip >= 1) {
            // Inner product value is positive.
            innerProductValue = score - 1;
        } else {
            // Inner product value is negative.
            innerProductValue = 1 - 1 / score;
        }

        System.out.println("Yes MOS (e.g. LuceneOnFaiss): final score=" + cosine(innerProductValue));
    }
}

Result:
No MOS (e.g. Faiss C++): final score=0.5
Yes MOS (e.g. LuceneOnFaiss): final score=0.5
No MOS (e.g. Faiss C++): final score=0.0
Yes MOS (e.g. LuceneOnFaiss): final score=0.0
No MOS (e.g. Faiss C++): final score=NaN
Yes MOS (e.g. LuceneOnFaiss): final score=NaN
```

Comment (0ctopus13prime) [reply src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java:84]: Since both are getting the same final score values with all the extreme values with the converting logic, I think it would be safe to keep it as it is?
Do you have thoughts on it?

File: src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java (hunk 4 above)
Comment (0ctopus13prime) [reply src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java:95]: `scoreDoc.score` is the result of MAX_IP conversion as shown below:

```
public static float scaleMaxInnerProductScore(float vectorDotProductSimilarity) {
    return vectorDotProductSimilarity < 0.0F ? 1.0F / (1.0F + -1.0F * vectorDotProductSimilarity) : vectorDotProductSimilarity + 1.0F;
}
```

If the score is 0, that means it had very large negative number (one example can be -Inf).
And for -Inf, the result of this cosine score encoding would be 0, which is exactly the one that we would have acquired with when memory optimized search disabled.

```
printResults(0);

    public static void printResults(final float ip) {
        // No MOS
        System.out.println("No MOS (e.g. Faiss C++): final score=" + cosine(ip));

        final float score = scaleMaxInnerProductScore(ip);
        final float innerProductValue;
        if (ip >= 1) {
            // Inner product value is positive.
            innerProductValue = score - 1;
        } else {
            // Inner product value is negative.
            innerProductValue = 1 - 1 / score;
        }

        System.out.println("Yes MOS (e.g. LuceneOnFaiss): final score=" + cosine(innerProductValue));
    }

// Result
No MOS (e.g. Faiss C++): final score=0.0
Yes MOS (e.g. LuceneOnFaiss): final score=0.0
```

The goal is to ensure both (LuceneOnFaiss, Faiss C++) yield the same score values.

File: src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java (hunk 1 above)
Comment (Vikasht34) [reply src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java:37]: Sure , Just want to Make Sure that , if anything changes on Lucene side on 10.4 , Please avoid making parallel changes in KNN , with this we will always creating parallel version of code in KNN and Lucene.

File: src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java (hunk 2 above)
Comment (Vikasht34) [reply src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java:84]: It should be fine , my concern was if we have differnce score from MOS vs Non Mos , as long as we have same it should be fine.

File: src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java (hunk 3 above)
Comment (Vikasht34) [reply src/main/java/org/opensearch/knn/index/query/MemoryOptimizedSearchScoreConverter.java]: sure if you are 100% sure on above cases will be handle and no drift will ever happen , I am fine.

**Review by Vikasht34 — APPROVED** [review]
I am good with Changes

**Review by navneet1v — APPROVED** [review]
